src/app/clusters/groups-server/groups-server.cpp

```diff
@@ -232,141 +261,137 @@ bool emberAfGroupsClusterViewGroupCallback(app::CommandHandler * commandObj, con
-            status = emberGetBinding(i, &entry);
-            if ((status == EMBER_SUCCESS) && (entry.type == EMBER_MULTICAST_BINDING) && (entry.local == emberAfCurrentEndpoint()))
+            TLV::TLVType type;
+            ReturnErrorOnFailure(writer.StartContainer(
+                TLV::ContextTag(to_underlying(Groups::Commands::GetGroupMembershipResponse::Fields::kGroupList)),
+                TLV::kTLVType_Array, type));
+
+            GroupDataProvider::GroupMapping mapping;
+            size_t requestCount = 0;
+            size_t matchCount   = 0;
+            ReturnErrorOnFailure(mCommandData.groupList.ComputeSize(&requestCount));
+
+            emberAfGroupsClusterPrint("RX: GetGroupMembership [");
+            while (mIterator->Next(mapping))
             {
-                list[listLen]     = EMBER_LOW_BYTE(entry.groupId);
-                list[listLen + 1] = EMBER_HIGH_BYTE(entry.groupId);
-                listLen           = static_cast<uint8_t>(listLen + 2);
-                count++;
-            }
-        }
-    }
-    else
-    {
-        auto iter = groupList.begin();
-        while (iter.Next())
-        {
-            GroupId groupId = iter.GetValue();
-            emberAfGroupsClusterPrint(" 0x%02" PRIx16, groupId);
-            for (j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
-            {
-                EmberBindingTableEntry entry;
-                status = emberGetBinding(j, &entry);
-                if ((status == EMBER_SUCCESS) && (entry.type == EMBER_MULTICAST_BINDING))
+                auto iter = mCommandData.groupList.begin();
+                while (iter.Next())
                 {
-                    if (entry.local == emberAfCurrentEndpoint() && entry.groupId == groupId)
+                    if (mapping.group == iter.GetValue())
                     {
-                        list[listLen]     = EMBER_LOW_BYTE(groupId);
-                        list[listLen + 1] = EMBER_HIGH_BYTE(groupId);
-                        listLen           = static_cast<uint8_t>(listLen + 2);
-                        count++;
+                        ReturnErrorOnFailure(chip::app::DataModel::Encode(writer, TLV::AnonymousTag, mapping.group));
+                        matchCount++;
+                        emberAfGroupsClusterPrint(" 0x%02" PRIx16, mapping.group);
+                        break;
                     }
                 }
+                ReturnErrorOnFailure(iter.GetStatus());
             }
+            emberAfGroupsClusterPrintln("]");
+
+            // 1.3.6.3.1: If the GroupList field contains one or more group-id’s but does not contain any group
+            // of which the entity is a member, the entity SHALL only respond if the command is unicast
+            VerifyOrReturnError((0 == requestCount) || (matchCount > 0) ||
+                                    (emberAfCurrentCommand()->type == EMBER_INCOMING_UNICAST),
+                                CHIP_ERROR_INVALID_DATA_LIST);
```

Comment:
This doesn't really make sense in Matter...  In particular, erroring out here will not prevent a response from attempting to be sent.  What will prevent that is the IM code not actually sending responses to groupcast things. Please just take this whole statement out.